### DISCUSSION TITLE
Automatic synchronization between @TestInstance and @Isolation

### DISCUSSION THREAD

**Discussion by ikysil** [cdi-unit/cdi-unit#290] [General]
JUnit 5 executes individual tests either in isolation (by creating an instance of the test for each method), or using the same test instance for all methods.

See [Test Instance Lifecycle](https://junit.org/junit5/docs/current/user-guide/#writing-tests-test-instance-lifecycle) for details.

CDI Unit provides similar controls for the lifecycle of the Weld container (`@Isolation` annotation and `IsolationLevel` enum).

At the moment, these controls are not synchronized, so there are four combinations possible:
- [A] `@TestInstance(Lifecycle.PER_METHOD)` + `@Isolation(IsolationLevel.PER_METHOD)`
- [B] `@TestInstance(Lifecycle.PER_METHOD)` + `@Isolation(IsolationLevel.PER_CLASS)`
- [C] `@TestInstance(Lifecycle.PER_CLASS)` + `@Isolation(IsolationLevel.PER_METHOD)`
- [D] `@TestInstance(Lifecycle.PER_CLASS)` + `@Isolation(IsolationLevel.PER_CLASS)`

> Side note: TestNG runner and listener support variant A only, JUnit 4 runner supports A and B, JUnit 4 rule supports A only.

Let's discuss how useful it is to synchronize these controls - so that only A and D are possible. 

Don't we loose some flexibility? Are there any corner cases which make any of the combinations a no-go?

**Comment by ikysil** [comment]
FYI: https://github.com/cschabl/cdi-unit-junit5 implements synchronization by completely ignoring `@Isolation` config. It supports variants A and D.

**Reply by graben** [reply]
IMHO variants A and D are the only reasonable ones. Ignoring `@Isolation` might work now, but may cause issues in the future. I'm in favour of setting `@Isolation` according to  `@TestInstance` by cdi-unit.

**Comment by ikysil** [comment]
I've added a test to verify the behavior at PR https://github.com/cdi-unit/cdi-unit/pull/293.
See [TestInstanceIsolation](https://github.com/cdi-unit/cdi-unit/blob/feat-support-nested-tests/junit5/src/test/java/io/github/cdiunit/junit5/tests/TestInstanceIsolation.java#L34) there.

Variants A and D are working as expected.

Variant B works as expected.

Variant C leads to issues with accessing non-dependent managed beans

> org.jboss.weld.exceptions.IllegalStateException: WELD-000229: Contextual reference of Managed Bean [class io.github.cdiunit.junit5.tests.TestInstanceIsolation$ApplicationCounter] with qualifiers [@Any @Default] is not valid after container cdi-unit-f2a5a12a-e335-4b4d-9ee3-106a2bfe4ce0 shutdown



I'm going to reflect on it for a couple of days.

**Comment by ikysil** [comment]
Hi @cschabl ,

Could you please chime in  ?

**Reply by cschabl** [reply]
Hi @ikysil ,
I ignored `@Isolation` in https://github.com/cschabl/cdi-unit-junit5, because I didn't know it until now (or can't remember it).
I was coming from from JUnit 5 and wanted to use CDI-Unit in JUnit 5 tests. It felt natural to me to use JUnit 5 built-in test instance lifecycle mechanism to control the lifecycle of the weld container.

As far as I can tell, lifecycle control of the weld instance in cdi-unit-junit5 and in cdi-unit with `@Isolation` ist basically the same.
If so, I would consider Junit 5's `@TestInstance.Lifeycle` an alternative for cdi-unit's `@Isolation`.
is there a reasonable use case for supporting a combination of
`TestInstance.Lifecycle.PER_CLASS`and 'IsolationLevel.PER_METHOD`and vice versa?

**Reply by ikysil** [reply]
Variant C is the default when using TestNG as it is not re-creating test instances, all tests are executed against the same instance.

**Comment by ikysil** [comment]
After adding more tests (in https://github.com/cdi-unit/cdi-unit/pull/299) I see the problem with variant C
```
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@Isolation(IsolationLevel.PER_METHOD)
```

Top-level test instance is managed by Weld container which gets shut down after the first test method leads to errors accessing that instance

> org.jboss.weld.exceptions.IllegalStateException: WELD-000229: Contextual reference of Managed Bean [class io.github.cdiunit.junit5.tests.TestIsolationPerClassTestPerMethodWeld] with qualifiers [@Any @Default] is not valid after container cdi-unit-7d61767b-d7e6-4fa7-b242-a4334a3e5040 shutdown


I will change the implementation so that `@TestInstance(TestInstance.Lifecycle.PER_CLASS)` overrides `@Isolation`.

**Comment by ikysil** [comment]
See #316 for more developments.
